src/common/oclengine.cpp

```diff
@@ -52,12 +77,97 @@ void OCLEngine::SetDeviceContextPtrVector(std::vector<DeviceContextPtr> vec, Dev
 
-void OCLEngine::InitOCL()
+cl::Program OCLEngine::MakeProgram(
+    bool buildFromSource, cl::Program::Sources sources, std::string path, std::shared_ptr<OCLDeviceContext> devCntxt)
+{
+    FILE* clBinFile;
+    cl::Program program;
+    cl_int buildError = -1;
+    std::vector<int> binaryStatus;
+    if (!buildFromSource && (clBinFile = fopen(path.c_str(), "r"))) {
+        long lSize;
+
+        fseek(clBinFile, 0L, SEEK_END);
+        lSize = ftell(clBinFile);
+        rewind(clBinFile);
+
+        std::vector<unsigned char> buffer(lSize);
+        lSize = fread(&buffer[0], sizeof(unsigned char), lSize, clBinFile);
+        fclose(clBinFile);
+
+#if defined(__APPLE__)
+        program = cl::Program(devCntxt->context, { devCntxt->device },
+            { std::pair<const void*, unsigned long>(&buffer[0], buffer.size()) }, &binaryStatus, &buildError);
+#else
+        program = cl::Program(devCntxt->context, { devCntxt->device }, { buffer }, &binaryStatus, &buildError);
+#endif
+
+        if ((buildError != CL_SUCCESS) || (binaryStatus[0] != CL_SUCCESS)) {
+            std::cout << "Binary error: " << buildError << ", " << binaryStatus[0] << " (Falling back to JIT.)"
+                      << std::endl;
+        } else {
+            std::cout << "Loaded binary." << std::endl;
+        }
+    }
+
+    // If, either, there are no cached binaries, or binary loading failed, then fall back to JIT.
+    if (buildError != CL_SUCCESS) {
+        program = cl::Program(devCntxt->context, sources);
+        std::cout << "Built JIT." << std::endl;
+    }
+
+    return program;
+}
+
+void OCLEngine::SaveBinary(cl::Program program, std::string path, std::string fileName)
+{
+    std::vector<size_t> clBinSizes = program.getInfo<CL_PROGRAM_BINARY_SIZES>();
+    size_t clBinSize = 0;
+    int clBinIndex = 0;
+    for (unsigned int i = 0; i < clBinSizes.size(); i++) {
+        if (clBinSizes[i] > 0) {
+            clBinSize = clBinSizes[i];
+            clBinIndex = i;
+            break;
+        }
+    }
+    std::cout << "Binary size:" << clBinSize << std::endl;
+
+    int err = mkdir(path.c_str(), 0700);
+    if (err != -1) {
+        std::cout << "Making directory: " << path << std::endl;
+    }
+
+    FILE* clBinFile = fopen((path + fileName).c_str(), "w");
+#if defined(__APPLE__)
+    std::vector<char*> clBinaries = program.getInfo<CL_PROGRAM_BINARIES>();
+    char* clBinary = clBinaries[clBinIndex];
+    fwrite(clBinary, clBinSize, sizeof(char), clBinFile);
+#else
+    std::vector<std::vector<unsigned char>> clBinaries = program.getInfo<CL_PROGRAM_BINARIES>();
+    std::vector<unsigned char> clBinary = clBinaries[clBinIndex];
+    fwrite(&clBinary[0], clBinSize, sizeof(unsigned char), clBinFile);
+#endif
+    fclose(clBinFile);
+}
+
+OCLInitResult OCLEngine::InitOCL(bool buildFromSource, bool saveBinaries, std::string home)
 {
+
+    if (home == "*") {
+        home = GetDefaultBinaryPath();
```

Comment:
Only do this if saveBinaries is set or !buildFromSource?

Reply:
I wouldn't. The parameter exists regardless. This is safer.